### PR TITLE
Convert Sensor Card to Typescript

### PR DESCRIPTION
**Pull request by zsarnett**
This is a beast of a Card. Im sure I messed something up...

### CODE REVIEW

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 1]
```diff
@@ -55,43 +59,80 @@ class HuiSensorCard extends EventsMixin(LitElement) {
...
+  protected shouldUpdate(changedProps: PropertyValues): boolean {
+    if (changedProps.has("_config")) {
+      return true;
+    }
+
+    const oldHass = changedProps.get("hass") as HomeAssistant | undefined;
+    if (oldHass) {
+      return (
+        oldHass.states[this._config!.entity] !==
+        this.hass!.states[this._config!.entity]
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: A sensor could have the same state but still have newer data points to display?

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 2]
```diff
@@ -195,46 +245,50 @@ class HuiSensorCard extends EventsMixin(LitElement) {
...
+      url += "&end_time=" + endTime.toISOString();
+    }
 
-  getCardSize() {
-    return 3;
+    return this.hass!.callApi("GET", url);
   }
 
-  _style() {
+  private renderStyle(): TemplateResult {
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: Can we move this to be right under the `render` function?

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 3]
```diff
@@ -100,38 +141,37 @@ class HuiSensorCard extends EventsMixin(LitElement) {
...
-
-  _computeIcon(item) {
-    return this._config.icon || stateIcon(item);
-  }
-
-  _computeName(item) {
-    return this._config.name || computeStateName(item);
-  }
+  private _handleClick(): void {
+    if (!this._config) {
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: When would `_config` not exist? Nothing should be rendered in that case.

File: src/panels/lovelace/common/create-card-element.js [hunk 4]
```diff
@@ -19,7 +19,7 @@ import "../cards/hui-picture-elements-card";
 import "../cards/hui-picture-entity-card";
 import "../cards/hui-picture-glance-card";
 import "../cards/hui-plant-status-card";
-import "../cards/hui-sensor-card";
+import "../cards/hui-sensor-card.ts";
```
Comment (iantrich) [inline src/panels/lovelace/common/create-card-element.js]: Should be no need to include the extension here

File: src/panels/lovelace/cards/hui-sensor-card.ts (hunk 2 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-sensor-card.ts]: Balloob Advised its the least import so put it at the end

File: src/panels/lovelace/common/create-card-element.js (hunk 4 above)
Comment (zsarnett) [reply src/panels/lovelace/common/create-card-element.js]: I thought that too. Ill try it again but it was broken for me

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 5]
```diff
@@ -1,52 +1,55 @@
...
+  detail?: number;
+  hours_to_show?: number;
+}
 
-  static get properties() {
+class HuiSensorCard extends LitElement implements LovelaceCard {
+  public hass?: HomeAssistant;
+  private _config?: Config;
+
+  static get properties(): PropertyDeclarations {
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: We want to render on updates to `hass`, don't we now that you've removed `shouldUpdate`?

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 6]
```diff
@@ -55,43 +58,65 @@ class HuiSensorCard extends EventsMixin(LitElement) {
...
 
-  render({ _entity, _line } = this) {
+  protected render(): TemplateResult {
+    if (!this._config || !this.hass) {
+      return html``;
+    }
+
+    const stateObj = this.hass.states[this._config.entity];
+    if (!stateObj) {
+      return html``;
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: Empty or error? I think rendering an error would be better.

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 7]
```diff
@@ -55,43 +58,65 @@ class HuiSensorCard extends EventsMixin(LitElement) {
...
+                            fill="none"
+                            stroke="var(--accent-color)"
+                            stroke-width="5"
+                            stroke-linecap="round"
+                            stroke-linejoin="round"
+                          />
+                        </svg>
+                      `;
+                    }),
+                    html``
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: An error for no `unit_of_measurement` could save users time in debugging here as well. My thinking is, the more clear errors we can show, the less questions in chat and issues in git :)

File: src/panels/lovelace/cards/hui-sensor-card.ts (hunk 5 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-sensor-card.ts]: Good catch

File: src/panels/lovelace/cards/hui-sensor-card.ts (hunk 7 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-sensor-card.ts]: If we were to do that then we would want to not default the graph. Meaning only show the graph if `config.graph === "line"` 

Dont want an error by default

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+    fireEvent(this, "hass-more-info", { entityId: this._config!.entity });
+  }
+
+  private _coordinates(
+    history: any,
+    hours: number,
+    width: number,
+    detail: number
+  ): number[][] {
+    history = history.filter((item) => !Number.isNaN(Number(item.state)));
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: We should first convert all items to numbers, then check for nan. Now you're converting to number again on line 147 and 148!

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+    const min = Math.min.apply(Math, history.map((item) => Number(item.state)));
+    const max = Math.max.apply(Math, history.map((item) => Number(item.state)));
+    const now = new Date().getTime();
+
+    const reduce = (res, item, point = false) => {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: I prefer no default param. Just add `false` to line 166

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+                    }</span
+                  >
+                </div>
+                <div class="graph">
+                  <div>
+                    ${
+                      this._config.graph !== "none" &&
+                      stateObj.attributes.unit_of_measurement
+                        ? until(
+                            this._getHistory().then((history) => {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: So eh, you're not caching this… So we're going to fetch the history on each change of `hass`, even if it wasn't even this entity that changed?

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+    const coords = this._coordinates(
+      stateHistory[0],
+      this._config!.hours_to_show!,
+      500,
+      this._config!.detail!
+    );
+    return this._getPath(coords);
+  }
+
+  private async _fetchRecent(startTime: Date, endTime: Date): Promise<{}> {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: We should move this to `src/data/history.ts`

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-sensor-card.ts]: Looks like it was already there

File: src/panels/lovelace/cards/hui-sensor-card.ts (hunk 7 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-sensor-card.ts]: image

I think by default it should not show the graph. Graph should be an opt in so I agree with you @iantrich 

Now will @balloob agree

**Comment by zsarnett** [comment]
I have a commit waiting to reset the date after pulling the history... I forgot to push.

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+
+  protected firstUpdated(): void {
+    this._date = new Date();
+  }
+
+  protected updated(changedProps: PropertyValues) {
+    const minute = 60000;
+    if (changedProps.has("_config")) {
+      this._getHistory();
+      this._date = new Date();
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: This should happen inside `_getHistory`

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+    }
+
+    if (Date.now() - this._date!.getTime() >= minute) {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: `else if` ?

File: src/panels/lovelace/cards/hui-sensor-card.ts [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+      path += ` ${Z[X]},${Z[Y]}`;
+      path += ` Q${next[X]},${next[Y]}`;
+      last = next;
+    }
+
+    path += ` ${next[X]},${next[Y]}`;
+    return path;
+  }
+
+  private _midPoint(
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-sensor-card.ts]: These kind of methods that do not use `this` can also be moved outside of the class, just a function above the class definition. 

```ts
const midPoint = (…) => { }
```

**Review by balloob — DISMISSED** [review]

**Comment by zsarnett** [comment]
NOt sure how to use the function Balloob was describing in Discord from the history element. Something to with Cache I believe.

**Review by bramkragten — APPROVED** [review]
Adding caching can be done in a new pr, it didn't have it before right?

**Pull request merged**